File: bundle-plugin/src/test/scala/com/yahoo/container/plugin/classanalysis/AnalyzeMethodBodyTest.scala
// Copyright 2016 Yahoo Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.classanalysis

import org.scalatest.junit.{JUnitSuite, ShouldMatchersForJUnit}
import sampleclasses._
import TestUtilities._
import org.junit.Test
import java.io.PrintStream

/**
 * Tests that classes used in method bodies are included in the imports list.
 * @author  tonytv
 */
class AnalyzeMethodBodyTest extends JUnitSuite with ShouldMatchersForJUnit {
  @Test def require_that_class_of_locals_are_included() {
    analyzeClass[Methods].referencedClasses should contain(name[Base])
  }

  @Test def require_that_class_of_locals_in_static_method_are_included() {
    analyzeClass[Methods].referencedClasses should contain(name[Derived])
  }

  @Test def require_that_field_references_are_included() {
    analyzeClass[Methods].referencedClasses should (contain (name[java.util.List[_]]) and contain (name[Fields]))
  }

  @Test def require_that_class_owning_field_is_included() {
    analyzeClass[Methods].referencedClasses should contain (name[System])
  }

  @Test def require_that_class_containing_method_is_included() {
    analyzeClass[Methods].referencedClasses should contain (name[PrintStream])
  }

  @Test def require_that_element_of_new_multidimensional_array_is_included() {
    analyzeClass[Methods].referencedClasses should contain (name[Interface1])
  }

  @Test def require_that_basic_arrays_are_not_included() {
    analyzeClass[Methods].referencedClasses should not (contain("int[]"))
  }

  @Test def require_that_container_generic_parameters_are_included() {
    analyzeClass[Methods].referencedClasses should contain(name[Dummy])
  }

  @Test def require_that_class_owning_method_handler_is_included() {
    analyzeClass[Methods].referencedClasses should contain(name[ClassWithMethod])
  }
}

File: bundle-plugin/src/test/java/com/yahoo/container/plugin/classanalysis/AnalyzeMethodBodyTest.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.classanalysis;

import com.yahoo.container.plugin.classanalysis.sampleclasses.Base;
import com.yahoo.container.plugin.classanalysis.sampleclasses.ClassWithMethod;
import com.yahoo.container.plugin.classanalysis.sampleclasses.Derived;
import com.yahoo.container.plugin.classanalysis.sampleclasses.Dummy;
import com.yahoo.container.plugin.classanalysis.sampleclasses.Fields;
import com.yahoo.container.plugin.classanalysis.sampleclasses.Interface1;
import com.yahoo.container.plugin.classanalysis.sampleclasses.Interface3;
import com.yahoo.container.plugin.classanalysis.sampleclasses.Methods;
import org.junit.jupiter.api.Test;

import java.io.PrintStream;
import java.util.List;

import static com.yahoo.container.plugin.classanalysis.TestUtilities.analyzeClass;
import static com.yahoo.container.plugin.classanalysis.TestUtilities.name;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Tests that classes used in method bodies are included in the imports list.
 *
 * @author Tony Vaagenes
 */
public class AnalyzeMethodBodyTest {

    @Test
    void class_of_locals_are_included() {
        assertTrue(analyzeClass(Methods.class).getReferencedClasses().contains(name(Base.class)));
    }

    @Test
    void class_of_locals_in_static_method_are_included() {
        assertTrue(analyzeClass(Methods.class).getReferencedClasses().contains(name(Derived.class)));
    }

    @Test
    void field_references_are_included() {
        assertTrue(analyzeClass(Methods.class).getReferencedClasses().containsAll(List.of(name(java.util.List.class), name(Fields.class))));
    }

    @Test
    void class_owning_field_is_included() {
        assertTrue(analyzeClass(Methods.class).getReferencedClasses().contains(name(System.class)));
    }

    @Test
    void class_containing_method_is_included() {
        assertTrue(analyzeClass(Methods.class).getReferencedClasses().contains(name(PrintStream.class)));
    }

    @Test
    void element_of_new_multidimensional_array_is_included() {
        assertTrue(analyzeClass(Methods.class).getReferencedClasses().contains(name(Interface1.class)));
    }

    @Test
    void basic_arrays_are_not_included() {
        assertFalse(analyzeClass(Methods.class).getReferencedClasses().contains("int[]"));
    }

    @Test
    void container_generic_parameters_are_included() {
        assertTrue(analyzeClass(Methods.class).getReferencedClasses().contains(name(Dummy.class)));
    }

    @Test
    void functional_interface_usage_is_included() {
        assertTrue(analyzeClass(Methods.class).getReferencedClasses().contains(name(Interface3.class)));
    }

    @Test
    void class_owning_method_handler_is_included() {
        assertTrue(analyzeClass(Methods.class).getReferencedClasses().contains(name(ClassWithMethod.class)));
    }
}
